lib/Sema/CSDiagnostics.cpp

```diff
@@ -2989,7 +2989,14 @@ bool TupleContextualFailure::diagnoseAsError() {
     diagnostic = diag::tuple_types_not_convertible_nelts;
   else if ((purpose == CTP_Initialization) && !getContextualType(getAnchor()))
     diagnostic = diag::tuple_types_not_convertible;
-  else if (auto diag = getDiagnosticFor(purpose, getToType()))
+  else if (purpose == CTP_AssignSource) {
```

Comment:
A better place to adjust this would be - `getStructuralTypeContext` which is responsible for computing `purpose` for this diagnostic.

Reply:
Ah makes sense. Fixing :)